src/main/java/io/anserini/collection/NewYorkTimesCollection.java

```diff
@@ -31,35 +31,40 @@
  */
 public class NewYorkTimesCollection extends Collection<NewYorkTimesDocument> {
   public class FileSegment extends Collection<NewYorkTimesDocument>.FileSegment {
-    private String fileName;
+    // We're creating a parser for each file, just to parse a single document, which is
+    // very inefficient. However, the parser is not thread safe, so this is our only option.
+    private final NewYorkTimesDocument.Parser parser = new NewYorkTimesDocument.Parser();
+
+    // Each file segment only has one file, boolean to keep track if it's been read.
+    private boolean docRead = false;
 
     protected FileSegment(Path path) throws IOException {
-      this.path = path;
-      this.fileName = path.toString();
+      super.path = path;
     }
 
     @Override
     public void close() throws IOException {
-      atEOF = false;
+      docRead = true;
+      super.close();
     }
 
     @Override
     public boolean hasNext() {
-      return !atEOF;
+      return !docRead;
     }
 
     @Override
     public NewYorkTimesDocument next() {
-      NewYorkTimesDocument doc = new NewYorkTimesDocument(new File(fileName));
-      atEOF = true;
+      if (docRead) return null;
```

Comment:
@Peilin-Yang would you prefer we return null here or throw a `NoSuchElementException`?

Reply:
`NoSuchElementException` is better.